AF-1871: Adding Clob support

jesuino opened the pull request:
Merge with https://github.com/kiegroup/appformer/pull/660

jesuino commented:
Would you please review, @ederign ?

To test it on kie server you must build kie server and register a query on a table that has a Clob column. Before this fix the column will not be in the results and will be filtered. After this fix the Clob column content should be available as String.

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
The test error `java.awt.AWTError: Can't connect to X11 window server using ':56' as the value of the DISPLAY variable.` does not seem related and it is not failing locally. Let me fire full downstream build again to make sure that everything is fine

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
Jenkins execute full downstream build

dgutierr commented:
@jesuino Can you please make sure the dataset SQL test suite [1] runs smoothly on the different DB supported? 

Notice that different JDBC api implementation might lead to inconsistencies and different behaviours when dealing with different DBs, so we need to make sure the CLOB support work all across the board.

Also a different issue is the "dashbuilder-dataset-sql-tests" shall be renamed as "kie-soup-dataset-sql-tests" and moved under the kie-soup repo. This wasn't done in the past because someone forgot to and I believe this might be the perfect time to do it. @ederign @porcelli wdyt?

[1] https://github.com/kiegroup/appformer/tree/master/dashbuilder/dashbuilder-backend/dashbuilder-dataset-sql-tests

jesuino commented:
> @jesuino @domhanak Can you please make sure the dataset SQL test suite [1] runs smoothly on the different DB supported?

It seems to work with DB2. I have a test error which is just local:

~~~
DB2Test.testAll:29->SQLTestSuite.testAll:45 expected:<143[8]> but was:<143[7]>
~~~

The same happens with H2 - it is possibly something wrong on my installation because we confirmed that such test failure does not happen in jenkins builds. I can go ahead and test with other databases as well. Please @domhanak let me know if you are also testing. 

> Notice that different JDBC api implementation might lead to inconsistencies and different behaviours when dealing with different DBs, so we need to make sure the CLOB support work all across the board.
> 
> Also a different issue is the "dashbuilder-dataset-sql-tests" shall be renamed as "kie-soup-dataset-sql-tests" and moved under the kie-soup repo. This wasn't done in the past because someone forgot to and I believe this might be the perfect time to do it. @ederign @porcelli wdyt?
> 
> [1] https://github.com/kiegroup/appformer/tree/master/dashbuilder/dashbuilder-backend/dashbuilder-dataset-sql-tests

Kindly confirm that it should be done and I can work on it. The version also needs to be updated, each tests suits referes to 0.7.0-SNAPSHOT, I think it is not being updated regularly.

jesuino commented:
Hello @dgutierr  - I moved dashbuilder-dataset-sql-tests to kie-soup and renamed it 
see #1ea44c5 - I also removed it from appformer, see:  https://github.com/kiegroup/appformer/pull/660

I also ran tests agains H2, DB2 and MariaDB - it was all fine - I can continue but first I would have to include some CLOB field test.

ederign commented:
@dgutierr would you mind to take a look on this when you have a chance?

ederign commented:
please jenkins execute full downstream build

jesuino commented on kie-soup-dataset/pom.xml:
```diff
@@ -38,6 +38,7 @@
     <module>kie-soup-dataset-shared</module>
     <module>kie-soup-dataset-core</module>
     <module>kie-soup-dataset-sql</module>
+    <module>kie-soup-dataset-sql-tests</module>
```
hello @dgutierr I understand what you mean, but if you check pom.xml for dashbuilder-dataset-sql-tests all the sub projects are commented with exactly what you said! :) 


https://github.com/kiegroup/appformer/blob/master/dashbuilder/dashbuilder-backend/dashbuilder-dataset-sql-tests/pom.xml

domhanak replied:
@dgutierr @jesuino We should prepare them to be part of the maven uild and test can be run with H2 by default as we already do for jbpm for example.

domhanak commented:
@jesuino imho having a submodule for each DB is an overkill and we should follow the pattern that jbpm or drools is following.

domhanak commented:
@jesuino check this and also the module out, imho we should do it like this - https://github.com/kiegroup/jbpm/blob/master/jbpm-persistence/jbpm-persistence-jpa/src/test/java/org/jbpm/persistence/util/PersistenceUtil.java

jesuino commented:
> Ok, so for this to be merged we need to have the same DB setup for integration tests as in drools and jbpm. Note that after this is done, it will also take some time to integrate it with QE automation.

Thanks for the feedback, working on it!

jesuino commented:
Hello @dgutierr @ederign  - just to clarify what Dominik is asking (please @domhanak correct me if I am wrong) 

- Instead having a project for each database type we should use a generic class that can create datasources of any type;
- To do so Dominik suggested [1] - however, kie-soup-datasource is not in the persistence layer, it handles directly the datasources
- In conclusion we actually need DataSourceFactory - and instad having tests for each db, we will have one test that will use DataSourceFactory to create datasources and, if no datasource setting is provided, we will use H2. This way QE can include kie-soup-datasource-sql-tests in their automation.

Next commit will follow with a possible solution.

[1] https://github.com/kiegroup/jbpm/blob/master/jbpm-persistence/jbpm-persistence-jpa/src/test/java/org/jbpm/persistence/util/PersistenceUtil.java

jesuino commented:
@domhanak Hello Dominik, there's a big issue about using the tests used y jBPM. I would have to move kie-test-all to kie-soup, the impact is that a few projects depends on kie-test-util, so I would have to make a major change or at least move kie-test-util to another BOM (currently it is in drools BOM) and make sure that all projects that refers to it will import the correct bom (perhaps kie-bom).

Do you agree with this?

Notice since kie-test-all is in drools I can't import it in kie-soup.

Thanks.

domhanak commented:
Couldn't we just reuse the setup of Database here? I am not sure I understand now, but what I meant is to have the setup for databases same as in jbpm or drools and run `kie-soup-datatset-sql`  tests using that DB setup. Imho there is no need to include another module,  we can probably create our own logic here - reusing the 
`import org.kie.test.util.db.DataSourceFactory;`
`import org.kie.test.util.db.PoolingDataSourceWrapper`

domhanak commented:
Update: We can probably get away with easier setup. Imho porting it as you mentioned might not be needed for the community tests, but I might be wrong. Let's discuss

domhanak commented on kie-soup-dataset/kie-soup-dataset-sql-tests/kie-soup-dataset-sql-tests-db2/src/test/java/org/dashbuilder/dataprovider/sql/DB2TestSettings.java:
```diff
@@ -0,0 +1,56 @@
+                String port = connectionSettings.getProperty("port");
+                String user = connectionSettings.getProperty("user");
+                String password = connectionSettings.getProperty("password");
+
```
Let's provide `datasource.properties` ad `filtered-resources`

`maxPoolSize=${maven.btm.maxPoolSize}
allowLocalTransactions=true
# JDBC/Database properties that are set in the maven pom
#
#   the below variable names (i.e. "${maven.datasource.classname}) are
# automagically replaced with their values (defined in the pom.xml)
# because of the fact that <filtering> is set to true in for the
# src/test/resources directory in the pom.
#
className=${maven.datasource.classname}
driverClassName=${maven.jdbc.driver.class}
user=${maven.jdbc.username}
password=${maven.jdbc.password}
url=${maven.jdbc.url}
serverName=${maven.jdbc.db.server}
portNumber=${maven.jdbc.db.port}
databaseName=${maven.jdbc.db.name}
defaultSchema=${maven.jdbc.schema}
dbBaseDir=${project.build.directory}
makeBaseDb=false
testMarshalling=false`

domhanak replied:
by default these should use H2

jesuino commented:
Hello Dominik,

Thanks for your feedback. I took parts of kie-test-util to create a generic test that could run with different databases. Could you please perform your tests with this new project and let me know if you see any issue? I will improve it tomorrow, just wanted to make sure that I am on the right direction.

Thanks!

See --> https://github.com/kiegroup/kie-soup/pull/83/commits/60ece5e6b947fddc6da874aad08705ca1098b32f

domhanak commented:
@jesuino Build fails on
```[ERROR]   
04:55:32 [ERROR]   The project org.kie.soup:kie-soup-dataset-sql-tests:[unknown-version] (/home/jenkins/workspace/KIE/master/pullrequest/kie-soup-pullrequests/kie-soup-dataset/kie-soup-dataset-sql-tests/pom.xml) has 1 error
04:55:32 [ERROR]     Non-resolvable parent POM for org.kie.soup:kie-soup-dataset-sql-tests:[unknown-version]: Could not find artifact org.kie.soup:kie-soup-dataset:pom:7.19.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 5, column 11 -> [Help 2]

jesuino commented:
jenkis retest this

jesuino commented:
@domhanak oops forgot to update a version, it should be fine now

jesuino commented on kie-soup-dataset/kie-soup-dataset-sql-tests/pom.xml:
```diff
@@ -0,0 +1,81 @@
+  <modules>
+    <module>kie-soup-dataset-sql-tests-generic</module>
+  <!--
+      Every test module has deps against a JDBC driver and a server instance.
```
@domhanak You mean the parent project? Probably I could move the code from kie-soup-dataset-sql-generic to kie-soup-dataset-sql-tests and avoid a parent project?

jesuino commented:
@domhanak thanks again for your review. I modified the tests and followed your suggestion and added a better README.md.

Let me know if it needs more improvements.

domhanak commented on kie-soup-dataset/kie-soup-dataset-sql/src/test/java/org/dashbuilder/dataprovider/sql/SQLColumnsTypeTest.java:
```diff
@@ -0,0 +1,76 @@
+        SQLDataSetDef def = new SQLDataSetDef();
+        def.setDbTable(CLOB_TABLE);
+        
+        def.setDataSource("jdbc:h2:mem:test");
```
Is this going to work with different datasources?

jesuino replied:
@domhanak It should work with multiple datasources, but this is the unit test for CLOB. Do you want to include a CLOB test in the kie-soup-datasource-sql-tests ?

domhanak replied:
is that feasible? Could be nice to cover it there too.

jesuino replied:
I just included it in the tests ran by kie-soup-dataset-sql-tests - so it should run for all databases - I found an issue with MySQL: in MySQL CLOB isn't a direct type, instead it is used behind some text types - hence the create table command had to be adjusted. For most databases it will run correctly (noticed it works with Oracle and DB2, for example) - but if it fails let me know and I will adjust the SQL - I think MySQL/MariaDB were the only exception :-) 

Let me know if you see any issue.

domhanak commented:
jenkins execute full downwstream build

domhanak commented:
Roses are red,
violets are blue,
Jenkins execute full downstream build!

ohonchar commented:
During testing PostgresBD....
_Result_:
`ERROR: Caused by: java.sql.SQLException: Cannot create PoolableConnectionFactory (FATAL: sorry, too many clients already)`
_Steps to reproduce:_
1. `docker run -d -p 5432:5432 --name postgresql -e POSTGRES_PASSWORD=pass -d postgres`
2. `docker exec -it postgresql psql -U postgres`
3. `CREATE DATABASE test WITH OWNER = postgres ENCODING = 'UTF8' LC_COLLATE = 'en_US.utf8' LC_CTYPE = 'en_US.utf8' TABLESPACE = pg_default CONNECTION LIMIT = -1;`
4. `mvn clean verify -Dmaven.datasource.classname=org.postgresql.xa.PGXADataSource -Dmaven.jdbc.driver.class=org.postgresql.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=5432 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/ohonchar/.m2/repository/org/postgresql/postgresql/42.2.5/postgresql-42.2.5.jar -Dmaven.jdbc.username=postgres -Dmaven.jdbc.password=pass -Dmaven.jdbc.url=jdbc:postgresql://localhost/test -Dmaven.jdbc.schema=public`

ohonchar commented:
During testing MySQLDB....
_Result:_
`[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.419 s <<< FAILURE! - in org.dashbuilder.dataprovider.sql.SQLDataSetTest
[ERROR] testAll(org.dashbuilder.dataprovider.sql.SQLDataSetTest)  Time elapsed: 1.283 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<"12/11/15 1[2]:00"> but was:<"12/11/15 1[4]:00">`

_Steps to reproduce:_
1. `docker run -d -p 3306:3306 --name=mysqldb --env="MYSQL_ROOT_PASSWORD=pass" mysql`
2. `docker exec -it mysqldb bash`
3. `mysql> mysql -u root -p`
4. `create database test;`
5. `use test;`
6. `mvn clean verify -Dmaven.datasource.classname=com.mysql.cj.jdbc.MysqlXADataSource -Dmaven.jdbc.driver.class=com.mysql.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=3306 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/ohonchar/.m2/repository/mysql/mysql-connector-java/8.0.15/mysql-connector-java-8.0.15.jar -Dmaven.jdbc.username=root -Dmaven.jdbc.password=pass -Dmaven.jdbc.url=jdbc:mysql://localhost/test -Dmaven.jdbc.schema=public`

domhanak commented:
@jesuino Please resolve the conflicts so we can get the WAR from FDB :)

jesuino commented:
@domhanak Removed the conflicts

@ohonchar Thanks for your feedback - I will first run with Postgres locally and try to reproduce that error - later I will check if I can reproduce the same test failure with MySQL - when I did MySQL I only tested CLOB support, didn't test everything because I saw unrelated test failures! But I am glad that MySQL at least didn't throw an error :)

jesuino commented:
@ohonchar 

I could not reproduce the following error:
~~~
ERROR: Caused by: java.sql.SQLException: Cannot create PoolableConnectionFactory (FATAL: sorry, too many clients already)
~~~
However, I did found an issue with tests on postgres - clob is not supported, hence I added a specific create table command for postgres. The only difference from your test is that I used a postgres not running locally and the version of my driver is slightly different:

~~~
mvn clean verify -Dmaven.datasource.classname=org.postgresql.xa.PGXADataSource -Dmaven.jdbc.driver.class=org.postgresql.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=5432 -Dmaven.jdbc.db.server=192.168.0.10 -Dmaven.jdbc.driver.jar=/home/wsiqueir/.m2/repository/org/postgresql/postgresql/42.2.2/postgresql-42.2.2.jar -Dmaven.jdbc.username=postgres -Dmaven.jdbc.password=pass -Dmaven.jdbc.url=jdbc:postgresql://192.168.0.10/test -Dmaven.jdbc.schema=public
~~~

The error you face seems to be due the default number of connections in postgres, that could be increased, please try a bigger value for MaxConnections in Postgres side.

Regarding MySQL, it was working when *I was running from Eclipse*, but now the same test fails here. It failed with the following message on your side:
~~~
testAll(org.dashbuilder.dataprovider.sql.SQLDataSetTest) Time elapsed: 1.283 s <<< FAILURE! org.junit.ComparisonFailure: expected:<"12/11/15 1[2]:00"> but was:<"12/11/15 1[4]:00">
~~~

This is what I got here:

~~~
[ERROR] testAll(org.dashbuilder.dataprovider.sql.SQLDataSetTest)  Time elapsed: 0.897 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<"12/11/15 1[2]:00"> but was:<"12/11/15 1[3]:00">
~~~

I think this may be a timezone issue. In MySQL the date is create at 12:00 but Java adjusts the timezone - probably due the use of SimpleDateFormat. Notice the same test still works with I run from Eclipse or try to debug using `-Dmaven.surefire.debug `.

jesuino commented:
![image](https://user-images.githubusercontent.com/359820/55241691-61814580-521a-11e9-8e69-ab2722a0ad28.png)

jesuino commented:
@ohonchar ok, the issue with MySQL seems to be the driver version. It works with driver version 5.x, for example, this works:

~~~
mvn clean verify -Dmaven.datasource.classname=com.mysql.jdbc.jdbc2.optional.MysqlXADataSource -Dmaven.jdbc.driver.class=com.mysql.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=3306 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/wsiqueir/Downloads/mysql-connector-java-5.1.47.jar -Dmaven.jdbc.username=root -Dmaven.jdbc.url=jdbc:mysql://localhost/test -Dmaven.jdbc.schema=public
~~~
But when I change the driver to 6.x it wont work anymore:
~~~
 mvn clean verify -Dmaven.datasource.classname=com.mysql.cj.jdbc.MysqlXADataSource -Dmaven.jdbc.driver.class=com.mysql.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=3306 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/wsiqueir/.m2/repository/mysql/mysql-connector-java/6.0.2/mysql-connector-java-6.0.2.jar -Dmaven.jdbc.username=root -Dmaven.jdbc.url=jdbc:mysql://localhost/test -Dmaven.jdbc.schema=public 
~~~

Could you please:

1) Increase the number of MaxConnections for Postgres
2) Use MySQL driver version 5.x 

Let me know please if it works.

We obviously have an issue with MySQL driver version 6+, we can open a separate ticket and work on that because it is not related to CLOB.

Thanks!

ohonchar commented:
> @ohonchar ok, the issue with MySQL seems to be the driver version. It works with driver version 5.x, for example, this works:
> 
> ```
> mvn clean verify -Dmaven.datasource.classname=com.mysql.jdbc.jdbc2.optional.MysqlXADataSource -Dmaven.jdbc.driver.class=com.mysql.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=3306 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/wsiqueir/Downloads/mysql-connector-java-5.1.47.jar -Dmaven.jdbc.username=root -Dmaven.jdbc.url=jdbc:mysql://localhost/test -Dmaven.jdbc.schema=public
> ```
> But when I change the driver to 6.x it wont work anymore:
> 
> ```
>  mvn clean verify -Dmaven.datasource.classname=com.mysql.cj.jdbc.MysqlXADataSource -Dmaven.jdbc.driver.class=com.mysql.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=3306 -Dmaven.jdbc.db.server=localhost -Dmaven.jdbc.driver.jar=/home/wsiqueir/.m2/repository/mysql/mysql-connector-java/6.0.2/mysql-connector-java-6.0.2.jar -Dmaven.jdbc.username=root -Dmaven.jdbc.url=jdbc:mysql://localhost/test -Dmaven.jdbc.schema=public 
> ```
> Could you please:
> 
> 1. Increase the number of MaxConnections for Postgres
> 2. Use MySQL driver version 5.x
> 
> Let me know please if it works.
> 
> We obviously have an issue with MySQL driver version 6+, we can open a separate ticket and work on that because it is not related to CLOB.
> 
> Thanks!

True, all two options that u  propose, resolving issues

domhanak commented:
Roses are red,
violets are blue,
Jenkins execute full downstream build!

jesuino commented:
@domhanak  - the test failure is unrelated IMO - will fire another full build

jesuino commented:
Jenkins execute full downstream build

domhanak commented:
@jesuino probably versions need updating to 7.21

jesuino commented:
Jenkins execute full downstream build

ohonchar commented:
@jesuino i've tried to test oracle db with clob type cell,
i've got last version from your pr and compiled it
after i did compilation of `kie-wb-distribution`
i copied `business-central.war` to eap and ran it
meanwhile i added configuration to `standalone.xml`
and what i've got:
in db i have:
![Selection_004](https://user-images.githubusercontent.com/38116191/56192688-9f160900-6037-11e9-9a7f-265c8525ce9f.png)
as u can see column2 has clob data type
and in BC i have:
![Selection_005](https://user-images.githubusercontent.com/38116191/56192714-a9d09e00-6037-11e9-9da9-9280246cb36c.png)
the query was `select * from "TEST"`
and u can see that it does not show `COLUMN2`

ohonchar commented:
Also, i've tested tests with oracle
this configs:
`mvn clean verify -Dmaven.datasource.classname=oracle.jdbc.xa.client.OracleXADataSource -Dmaven.jdbc.driver.class=oracle.jdbc.Driver -Dmaven.jdbc.db.name=test -Dmaven.jdbc.db.port=32769 -Dmaven.jdbc.db.server=127.0.0.1 -Dmaven.jdbc.driver.jar=/home/ohonchar/Projects/bmxs/BC/jboss-eap-installer/ojdbc8.jar -Dmaven.jdbc.username="sys as sysdba" -Dmaven.jdbc.password=Oradoc_db1 -Dmaven.jdbc.url=jdbc:oracle:thin:@127.0.0.1:32769:ORCLCDB -Dmaven.jdbc.schema=public`

and got this results:
[oracle-clob-errors.txt](https://github.com/kiegroup/kie-soup/files/3089328/oracle-clob-errors.txt)

ederign commented:
@ohonchar as soon as you are happy with this one, could you please also approves https://github.com/kiegroup/appformer/pull/660

ohonchar commented:
> @ohonchar as soon as you are happy with this one, could you please also approves [kiegroup/appformer#660](https://github.com/kiegroup/appformer/pull/660)

Sure, thnx

jesuino commented:
Hello @ohonchar 

Thanks again for your tests with Oracle Database.

You should be able to use clob columns in the datasource admin UI because it is just an interface for the underlying dataset APIs.

I did test with Oracle 12c and the CLOB test passes. The database and driver I used is the same:

********************************************************************************************
Database: Oracle Oracle Database 12c Enterprise Edition Release 12.2.0.1.0 - 64bit Production
Driver: Oracle JDBC driver 12.1.0.1.0
*********************************************************************************************

What I am wondering is if your user has access to create such table, although the error should be different if that was the case . If you login directly in the database are you able to run the following create table?

CREATE TABLE CLOB_TABLE (ID INTEGER PRIMARY KEY,CLOB_CL CLOB);

Thanks.

jesuino commented:
Hello @ohonchar 

For the unit tests the issue seemed to be the additional semicolon in the Create table for Oracle - Oracle does not seem to like it. Could you please pull commit a024695c95dcd0f0260a1feba0f37ae63bae98c4 and let me know if it works?

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
Hello @ohonchar

Regarding the UI: It works from the UI, see the following video. I used dashbuilder for a quick test, but tomorrow we should have a business central war from the Full build I am running - so I can test with business central

ohonchar commented:
@jesuino Hello, I can approve, that last commit works for me as well. UI and unit tests finished with success. 
Approve from my side.
Thnx for your effort William!

jesuino commented:
tests failure seems to be unrelated - would you please merge it, @ederign ?
